solutions/USACO feb20-clocktree-silver.mdx

```diff
@@ -0,0 +1,109 @@
+---
+id: usaco-1016
+source: USACO Silver 2020 February
+title: Clock Tree
+author: Melody Yu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_clocktree_silver_feb20.html)
+
+We can build a graph with the rooms as nodes and corridors as edges. Consider a simple case: there are 2 nodes in the graph and one edge.
+Bessie can start from one of the rooms, go to its neighbor, increase the clock, go back, and increase the starting room's clock.
+During this process of moving back and forth, notice that the difference between two rooms' clock readings may only change by 1.
+If the starting clock readings differ by one, Bessie can start from the bigger clock reading room and travel back and forth until both of the clocks point to 12.
+If the starting clock readings differ by more than one, it is impossible for Bessie to make both clocks point to 12.
+
+If there are more nodes, we can treat it similar to two node case by considering the two groups in a bipartite partition. Note that all trees are bipartite graphs, so this partition will always be successful.
+
+Every time Bessie visits a neighbor node, the sum of clocks of the group that Bessie visits increase by 1.
+Then Bessie visits another neighbor node, the sum of clocks of the second group increase by 1.
+As both groups increase clock by one, the difference between the sum of clocks for these two groups does not change.
+This process repeats until clocks in 1 or 2 groups reach all 12.
+
+We can check the initial sum difference between two groups (mod 12) to figure out the final clock numbers.
+
+1. If group0 is same as group2 (mod 12), Bessie can start from any node
+2. If group0 is 1 smaller than group1 (mod 12), Bessie starts from any node of group1
+3. if group1 is 1 smaller than group0 (mod 12), Bessie starts from any node of group0
+4. If group0 and group1 differ by more than 2 (mod 12), there's no way to get all 12
+
+## Video Solution
+
+<Youtube id="vMA54A2AT0I" />
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int N;
+vector<int> edges[100000];
+int A[100000];
+int sum0, sum1, nodes0, nodes1;
+
+void dfs(int i, int color, int par) {
+	// update color/sum
+	if (color == 0) {
+		nodes0++;
+		sum0 += A[i];
+	}
+	else {
+		nodes1++;
+		sum1 += A[i];
+	}
+	
+	for(int j = 0; j < edges[i].size(); j++) {
```

Comment:
also pretty minor but why don't you use a for-each loop here? (`for (int j: edges[i])`)